proposals/fixed-sized-buffers.md

```diff
@@ -29,42 +31,51 @@ public fixed DXGI_RGB GammaCurve[1025];
-    {
-        get;
-    }
+    public ref T this[int index] => ref (uint)index <= 1024u ?
+                                         ref RefAdd<T>(ref _e0, index):
+                                         throw new IndexOutOfRange();
 }
 ```
 
+Since such fixed-sized buffers no longer require use of `fixed`, it makes sense to allow any element type.  
+
+> NOTE: `fixed` will still be supported, but only if the element type is `blittable`
+
 ## Drawbacks
 [drawbacks]: #drawbacks
 
-Some additional syntax may be needed to ensure we don't break users that are currently using 'unsafe fixed-buffers'.
-* Given that the existing fixed-sized buffers only work with a selection of primitive types, it should be possible for the compiler to continue "just-working" if the user treats the fixed-buffer as a pointer.
-* Its probably worth extending the ability to work with fixed-sized buffers as pointers to all `blittable` types.
-
-In order to ensure that the safe fixed-sized buffer is resilient to future field-layout changes (and for it to work with non-blittable types), the compiler needs to explicitly layout each field of the array.
-* This can quickly become 'unwieldly' for large arrays
-* The JIT may not behave well with types that contain a large number of fields
- * We may be able to work with the JIT/Runtime to provide a special flag they can use (basically, it would give the contained type and the element count, then they just need to compute the size of the base struct and multiply, rather than actually checking each field).
+* There could be some challenges with backwards compatibility, but given that the existing fixed-sized buffers only work with a selection of primitive types, it should be possible for the compiler to continue "just-working" if the user treats the fixed-buffer as a pointer. 
+* Incompatible constructs may need to use slightly different `v2` encoding to hide the fields from old compiler.
+* Packing is not weel defined in IL spec for generic types. While the approach should work, we will be bordering on undocumnted behavior. We should make that documented and make sure other JITs like Mono have the same behavior.
+* Specifying a separate type for every length (an possibly another for `readonly` fields, if supported) will have impact on metadata. It will be bound by the number of arrays of different sizes in the given app. 
+* `ref` math is not formally verifiable (since it is unsafe). We will need to find a way to update verification rules to know that our use is ok.
 
 ## Alternatives
 [alternatives]: #alternatives
 
-Manually declare your structures
+Manually declare your structures and use unsafe code to construct indexers.
 
 ## Unresolved questions
 [unresolved]: #unresolved-questions
 
-Can/should this additionally be extended to stackalloc which provides a very similar mechanism.
+- should we allow `readonly`?  (with readonly indexer)
+- should we allow array initializers?
+- is `fixed` keyword necessary?
```

Comment:
Hmm, `fixed` seems pointless if the construct is safe.

Reply:
I think we need some syntax to disambiguate normal arrays from inline arrays. Although I'm not sure if `fixed` is the right way to do that 😄

Reply:
Well, the syntax is rather different already:
```C#
struct Foo {
    int[] array;
    int buffer[3];
}
```
I'm sure some will be confused at first but I doubt that this confusion will have negative long term consequences. In the worst case they'll think that copying such a struct will result in a new struct that references the old buffer :smile:

> Although I'm not sure if fixed is the right way to do that

`inline` perhaps? Not an existing keyword so perhaps problematic. or `struct`?
```C#
struct Foo {
    inline int buffer[3];
    struct int buffer[3]; // kind of communicates that we're really dealing with a hidden, anonymous struct?
}
```
Both seem a bit silly though.

Reply:
I would prefer `inline` over `struct`, as it seems a bit more clear. But you're right, the syntax is already different so maybe there won't be any issues